cmd/query/main.go

```diff
@@ -24,57 +24,88 @@ import (
-	if err != nil {
-		logger.Fatal("Failed to create span reader", zap.Error(err))
-	}
-	dependencyReader, err := storageBuild.NewDependencyReader()
-	if err != nil {
-		logger.Fatal("Failed to create dependency reader", zap.Error(err))
+	var command = &cobra.Command{
+		Use:   "jaeger-query",
+		Short: "Jaeger query is a service to access tracing data",
+		Long:  `Jaeger query is a service to access tracing data and host UI.`,
+		Run: func(cmd *cobra.Command, args []string) {
+			casOptions.InitFromViper(v)
+			queryOpts := &builder.QueryOptions{}
+			queryOpts.InitFromViper(v)
+			sFlags := &flags.SharedFlags{}
+			sFlags.InitFromViper(v)
+
+			metricsFactory := xkit.Wrap("jaeger-query", expvar.NewFactory(10))
+
+			storageBuild, err := builder.NewStorageBuilder(
+				sFlags.SpanStorage.Type,
+				sFlags.DependencyStorage.DataFrequency,
+				basicB.Options.LoggerOption(logger),
+				basicB.Options.MetricsFactoryOption(metricsFactory),
+				basicB.Options.CassandraOption(casOptions.GetPrimary()),
+			)
+			if err != nil {
+				logger.Fatal("Failed to init storage builder", zap.Error(err))
+			}
+			spanReader, err := storageBuild.NewSpanReader()
+			if err != nil {
+				logger.Fatal("Failed to create span reader", zap.Error(err))
+			}
+			dependencyReader, err := storageBuild.NewDependencyReader()
+			if err != nil {
+				logger.Fatal("Failed to create dependency reader", zap.Error(err))
+			}
+			rHandler := app.NewAPIHandler(
+				spanReader,
+				dependencyReader,
+				app.HandlerOptions.Prefix(queryOpts.QueryPrefix),
+				app.HandlerOptions.Logger(logger))
+			sHandler := app.NewStaticAssetsHandler(queryOpts.QueryStaticAssets)
+			r := mux.NewRouter()
+			rHandler.RegisterRoutes(r)
+			sHandler.RegisterRoutes(r)
+			portStr := ":" + strconv.Itoa(queryOpts.QueryPort)
+			recoveryHandler := recoveryhandler.NewRecoveryHandler(logger, true)
+			logger.Info("Starting jaeger-query HTTP server", zap.Int("port", queryOpts.QueryPort))
+			if err := http.ListenAndServe(portStr, recoveryHandler(r)); err != nil {
+				logger.Fatal("Could not launch service", zap.Error(err))
+			}
+		},
 	}
-	rHandler := app.NewAPIHandler(
-		spanReader,
-		dependencyReader,
-		app.HandlerOptions.Prefix(*builder.QueryPrefix),
-		app.HandlerOptions.Logger(logger))
-	sHandler := app.NewStaticAssetsHandler(*builder.QueryStaticAssets)
-	r := mux.NewRouter()
-	rHandler.RegisterRoutes(r)
-	sHandler.RegisterRoutes(r)
-	portStr := ":" + strconv.Itoa(*builder.QueryPort)
-	recoveryHandler := recoveryhandler.NewRecoveryHandler(logger, true)
-	logger.Info("Starting jaeger-query HTTP server", zap.Int("port", *builder.QueryPort))
-	if err := http.ListenAndServe(portStr, recoveryHandler(r)); err != nil {
-		logger.Fatal("Could not launch service", zap.Error(err))
+
+	flagSet := &flag.FlagSet{}
+	flags.AddFlags(flagSet)
+	casOptions.AddFlags(flagSet)
+	builder.AddFlags(flagSet)
+	command.PersistentFlags().AddGoFlagSet(flagSet)
+
+	v.BindPFlags(command.PersistentFlags())
```

Comment:
See at the collector and standalone main. I maybe like more collector because it does not brake command struct.